### PR TITLE
Annotate framework tests with where they are expected to skip in CI

### PR DESCRIPTION
**Pull request by jon-turney**
Replace the hard-coded list of 'may be skipped' framework tests in `skippable()` with annotations in `test.json` which record 'will be skipped in these specific CI jobs'.

The motivation here is that, at the moment, if something is dropped from any VM image our CI runs on, if it's permitted by `skippable()`, we'll just silently stop testing it.  This is particularly relevant where we don't control the VM image (e.g. on Windows), and has actually happened with boost on Windows (see the comments for commit 5c0cb0547bd52230faff322a33bfeac0eaaeea0d).

### CODE REVIEW

**Comment by jon-turney** [comment]
As it stands, this PR results in two different meanings of 'skip' (which could therefore use some better naming):

1. Test isn't run, because we 'know' it can't work in this enviroment
2. Test is run, and is expected to output 'MESON_SKIP_TEST'

Possible future work:

- [ ] Set MESON_CI_JOBNAME in all CI jobs
- [ ] Add test.json skip annotations for those jobs
- [ ] Make it an error if CI is set, but MESON_CI_JOBNAME isn't
- [ ] Extend to test categories other than framework?

File: docs/markdown/Contributing.md [hunk 1]
```diff
@@ -380,6 +378,30 @@ matched:
...
+#### skip_on_env
+
+The `skip_on_env` key can be used to specify a list of environment variables. If
+at least one environment variable in the `skip_on_env` list is present, the test
+is skipped.
+
+#### skip_on_jobname
+
+The `skip_on_jobname` key contains a list of strings. If the `MESON_CI_JOBNAME`
+environment variable is set, and any of them are contained in it, the test is
```
Comment (mensinda) [inline docs/markdown/Contributing.md:390]: This should probably make it clearer that only a substring has to be matched, because I missed that part initially.

File: run_project_tests.py [hunk 2]
```diff
@@ -264,22 +264,22 @@ def __lt__(self, other: object) -> bool:
...
+compile_commands:   T.List[str] = []
+clean_commands:     T.List[str] = []
+test_commands:      T.List[str] = []
+install_commands:   T.List[str] = []
+uninstall_commands: T.List[str] = []
 
-backend:      'Backend'
-backend_flags: T.List[str]
+backend:       Backend = Backend.ninja
+backend_flags: T.List[str] = []
```
Comment (mensinda) [inline run_project_tests.py:282]: Are these changes really required or automatically generated?

File: run_project_tests.py [hunk 3]
```diff
@@ -731,7 +730,37 @@ def force_regenerate() -> None:
...
+
+    # Test is expected to skip if os matches
+    if 'skip_on_os' in test_def:
+        for skip_os in test_def['skip_on_os']:
+            if skip_os.startswith('!'):
+                if mesonenv.machines.host.system != skip_os[1:]:
+                    skip_expected = True
+            else:
+                if mesonenv.machines.host.system == skip_os:
+                    skip_expected = True
```
Comment (mensinda) [inline run_project_tests.py:751]: Isn't using a full `Environment` way too overkill here? Alternatively, wouldn't it also be possible to create a "fake" Environment without a build dir, thus eliminating the need of an `TemporaryDirectoryWinProof`, directly here?

**Review by mensinda — COMMENTED** [review]
Overall, I really like this change. However, I have some minor questions.

File: run_project_tests.py (hunk 2 above)
Comment (jon-turney) [reply run_project_tests.py:282]: The initializers are needed to prevent flake8 from reporting `F821 undefined name` (which on reflection, I guess is a bug in flake8, since the name should be defined by type-annotated name, even without an initial value)

I'm not sure the current annotation for the `backend` global is actually right:  the string `backend` isn't the type `backend`?

Maybe I should just drop the flake8 fixes commit (since we don't seem to be running flake8/sider anymore)

File: docs/markdown/Contributing.md (hunk 1 above)
Comment (jon-turney) [reply docs/markdown/Contributing.md:390]: Yeah, I'll adjust that text.

I did initally think of making the jobnames a bit more structured (e.g. a series of '-' separated strings, and exactly matching one of those, so e.g. foo and nofoo were matchable), but that didn't seem to be needed yet.

File: run_project_tests.py (hunk 3 above)
Comment (jon-turney) [reply run_project_tests.py:751]: There's only a single Environment object made per detect_tests_to_run(), so it's not particularly heavyweight, and I think it's possible in future we might want to have some other tests in load_tests_json which require one.

(There's some refactoring which could be done to pass an Environment object around and down to stuff which detect_tests_to_run calls which needs an Environment, e.g. have_objc_compiler etc., which each make one at the moment.
The `with TemporaryDirectoryWinProof` then construct a `environment.Environment` stanza appears 4 times in this file now)

File: run_project_tests.py (hunk 2 above)
Comment (mensinda) [reply run_project_tests.py:282]: I would say that we drop the changes then.

> I'm not sure the current annotation for the backend global is actually right: the string 'backend' isn't the type backend?

The `'Backend'` is a valid type annotation. It is a string so that `Backend` does not have to be imported outside the `if T.TYPE_CHECKING:` block.

File: run_project_tests.py (hunk 3 above)
Comment (mensinda) [reply run_project_tests.py:751]: My main problem is that it is passed through a lot of functions just that don't do anything with it, just to be used here...

So, could you try if `environment.Environment(None, None, get_fake_options('/'))` (without a temp dir) also works, so that it can be directly constructed here? If it doesn't then I guess it is fine the way it currently is.

**Review by mensinda — APPROVED** [review]
LGTM and the OSX CI fails because of clock skew. Thanks for fixing/implementing this :)
